iothub/device/samples/PnpDeviceSamples/TemperatureController/Program.cs

```diff
@@ -0,0 +1,327 @@
+            await s_deviceClient.SetMethodHandlerAsync("thermostat1*getMaxMinReport", HandleMaxMinReportCommandAsync, Thermostat1);
+            await s_deviceClient.SetMethodHandlerAsync("thermostat2*getMaxMinReport", HandleMaxMinReportCommandAsync, Thermostat2);
+
+            s_logger.LogDebug($"Set handler to receive \"targetTemperature\" updates.");
+            await s_deviceClient.SetDesiredPropertyUpdateCallbackAsync(SetDesiredPropertyUpdateCallbackAsync, null);
+            s_desiredPropertyUpdateCallbacks.Add(Thermostat1, TargetTemperatureUpdateCallbackAsync);
+            s_desiredPropertyUpdateCallbacks.Add(Thermostat2, TargetTemperatureUpdateCallbackAsync);
+
+            await SendDeviceMemoryAsync();
+            await SendDeviceSerialNumberAsync();
+
+            bool temperatureReset = true;
+            s_maxTemp[Thermostat1] = 0d;
+            s_maxTemp[Thermostat2] = 0d;
+
+            await Task.Run(async () =>
+            {
+                while (true)
+                {
+                    if (temperatureReset)
+                    {
+                        // Generate a random value between 5.0°C and 45.0°C for the current temperature reading for each "Thermostat" component.
+                        s_temperature[Thermostat1] = Math.Round(s_random.NextDouble() * 40.0 + 5.0, 1);
+                        s_temperature[Thermostat2] = Math.Round(s_random.NextDouble() * 40.0 + 5.0, 1);
+                    }
+
+                    await SendTemperatureAsync(Thermostat1);
+                    await SendTemperatureAsync(Thermostat2);
+
+                    temperatureReset = s_temperature[Thermostat1] == 0 && s_temperature[Thermostat2] == 0;
+                    await Task.Delay(5 * 1000);
+                }
+            });
+        }
+
+        // Initialize the device client instance over Mqtt protocol (TCP, with fallback over Websocket), setting the ModelId into ClientOptions.
+        // This method also sets a connection status change callback, that will get triggered any time the device's connection status changes.
+        private static void InitializeDeviceClientAsync()
+        {
+            var options = new ClientOptions
+            {
+                ModelId = ModelId,
+            };
+            s_deviceClient = DeviceClient.CreateFromConnectionString(s_deviceConnectionString, TransportType.Mqtt, options);
+            s_deviceClient.SetConnectionStatusChangesHandler((status, reason) =>
+            {
+                s_logger.LogDebug($"Connection status change registered - status={status}, reason={reason}.");
+            });
+        }
+
+        // Send working set of device memory over telemetry.
+        private static async Task SendDeviceMemoryAsync()
+        {
+            string telemetryName = "workingSet";
+            long workingSet = Environment.WorkingSet / 1024;
+
+            Message msg = PnpHelper.CreateIothubMessageUtf8(telemetryName, JsonConvert.SerializeObject(workingSet));
+
```

Comment:
Have you aligned with Other SDK's on this PnPHelper design?

Reply:
This is in line with what C has: https://github.com/Azure/azure-iot-sdk-c/blob/jspaith_pnp_sample_on_deviceclient/iothub_client/samples/pnp/common/pnp_protocol_helpers.c
Node and python have similar logic, with the actual method invocation within the helper, instead of in the sample directly. I have synced up with them on this, and they will be fixing that.